Adding azure case in AnnotateScmSlugEntityProcessor

taraknath555 opened the pull request:
## Hey, I just made a Pull Request!

Set azure annotation dev.azure.com/project-repo in AnnotateScmSlugEntityProcessor to find the project and repo information for the repos that contains dev.azure.com in the url.

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

awanlin reviewed (changes requested):
Thanks for the contribution @taraknath555, this is a good idea but I think we should make sure to support both Cloud and Server versions of Azure DevOps. This should be achievable if you use `git-url-parse`, there are many examples in this repo. 

Also, I'm not sure if `backstage.io/managed-by-location` will work for everyone, we keep our YAML in a central repo so it would pull in the wrong project and repo for us. Could we use `backstage.io/source-location` instead?

Finally, if we could add some tests for this that would be really good!

cal5barton commented on plugins/azure-devops/README.md:
````diff
@@ -61,6 +61,12 @@ spec:
+backstage.io/managed-by-location: https://dev.azure.com/{org}/{project}/_git/{repo}?path=/.catalog/component.yml
+```
+
+if `dev.azure.com/project-repo` annotation is not provided and `dev.azure.com/project` not exits then to extract the project and repo information we can use `backstage.io/managed-by-location` annotation if location contains `dev.azure.com` in url to enable CICD tab.
````
This needs to be more clear. Maybe something like this?
```
If the annotation dev.azure.com/project-repo is absent and the dev.azure.com/project does not exist, we can derive the project and repository details by utilizing the backstage.io/managed-by-location annotation. This approach becomes active when the URL within the location contains dev.azure.com, allowing the activation of the CICD tab.
```

cal5barton commented:
> Thanks for the contribution @taraknath555, this is a good idea but I think we should make sure to support both Cloud and Server versions of Azure DevOps. This should be achievable if you use `git-url-parse`, there are many examples in this repo.
> 
> Also, I'm not sure if `backstage.io/managed-by-location` will work for everyone, we keep our YAML in a central repo so it would pull in the wrong project and repo for us. Could we use `backstage.io/source-location` instead?
> 
> Finally, if we could add some tests for this that would be really good!

It seems like this method would work to support both correct? https://github.com/backstage/backstage/blob/master/plugins/github-actions/src/components/getHostnameFromEntity.ts

It also seems like this method should be more of a core helper method. Is that something we should consider?

cal5barton commented on plugins/azure-devops/src/constants.ts:
```diff
@@ -18,4 +18,5 @@ export const AZURE_DEVOPS_BUILD_DEFINITION_ANNOTATION =
   'dev.azure.com/build-definition';
 export const AZURE_DEVOPS_PROJECT_ANNOTATION = 'dev.azure.com/project';
 export const AZURE_DEVOPS_REPO_ANNOTATION = 'dev.azure.com/project-repo';
+export const AZURE_DEVOPS_PROJECT_LOCATION = 'backstage.io/managed-by-location';
```
This won't be needed.

cal5barton commented on plugins/azure-devops/src/utils/isAzureLocation.ts:
```diff
@@ -0,0 +1,24 @@
+import { Entity } from '@backstage/catalog-model';
+import { AZURE_DEVOPS_PROJECT_LOCATION } from '../constants';
+
+export const isDevAzureLocation = (entity: Entity): boolean => {
```
If we end up making this method https://github.com/backstage/backstage/blob/master/plugins/github-actions/src/components/getHostnameFromEntity.ts a core helper method then this won't be needed, however, if not we should mirror the functionality there in this file.

awanlin commented:
Hi @taraknath555, I only recently learned about the existence of this processor but I'm think it would be much smarter to add this feature to the existing `AnnotateScmSlugEntityProcessor` as it then becomes opt-in for users and would support both Cloud and Server usage. 

To implement this all you would need to do is add the `dev.azure.com/project-repo` annotation here: https://github.com/backstage/backstage/blob/263e3846d8e0d2270800f7ddc224da902912f53e/plugins/catalog-backend/src/modules/core/AnnotateScmSlugEntityProcessor.ts#L29

And then update this switch with a case for `azure`: https://github.com/backstage/backstage/blob/263e3846d8e0d2270800f7ddc224da902912f53e/plugins/catalog-backend/src/modules/core/AnnotateScmSlugEntityProcessor.ts#L77

taraknath555 commented:
Hey @awanlin, can you please make me clear if only need to do the below change in this pull request?
![image](https://github.com/backstage/backstage/assets/42321546/88e707f8-49a1-4b8a-bf15-151f36337dcf)
 
If so the other part of the code that will make the requirement work is taken care by #19616 ?

awanlin commented:
Hi @taraknath555, yes, I would just make the changes I suggested to add getting the `dev.azure.com/project-repo` annotation set in the `AnnotateScmSlugEntityProcessor`. I don't want you to have to wait around for my changes to get merged in as they may take a while as they get reviewed. Just doing this will be a big win for the community 🚀 

Also, thanks for taking my various requests in stride, I really appreciate it 👍

taraknath555 commented:
> Hi @taraknath555, yes, I would just make the changes I suggested to add getting the `dev.azure.com/project-repo` annotation set in the `AnnotateScmSlugEntityProcessor`. I don't want you to have to wait around for my changes to get merged in as they may take a while as they get reviewed. Just doing this will be a big win for the community 🚀
> 
> Also, thanks for taking my various requests in stride, I really appreciate it 👍

Hey @awanlin, I updated the PR as per you suggested, Please take a look on this!

awanlin reviewed (changes requested):
Looks good to me but we should add a test for this in `AnnotateScmSlugEntityProcessor.test.ts`, please

taraknath555 commented:
> Looks good to me but we should add a test for this in `AnnotateScmSlugEntityProcessor.test.ts`, please

Done!

awanlin reviewed (approved):
Looks good to me, approving but if you can update the PR description as well with what was actually done would be great

taraknath555 commented:
> what was actually done would be great

Hey @awanlin i updated the PR title! Now, Where you want me to add the things that is actually done for this PR ?

taraknath555 commented:
> > what was actually done would be great
> 
> Hey @awanlin i updated the PR title! Now, Where you want me to add the things that is actually done for this PR ?
Hey @awanlin can you please suggest what else changes you required , i already updated the description/title of PR and when can we merge this PR ?

freben commented on .changeset/tiny-bulldogs-grow.md:
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/plugin-azure-devops': patch
```
```suggestion
'@backstage/plugin-catalog-backend': minor
```

since the change is in that package, and a feature addition

freben replied:
You still didn't change the package name.

freben reviewed (commented):
The catalog maintainers are the ones to help merge this. You'll have to wait for their review too.

awanlin commented:
Hi @taraknath555, sorry this fell off my radar. Can you also update the description of this PR? It still talks about the previous changes and has the image that isn't relevant anymore. You could simply replace that with: "Set azure annotation `dev.azure.com/project-repo` in `AnnotateScmSlugEntityProcessor` to find the project and repo information for the repos that contains `dev.azure.com` in the url"

As already mentioned I'm just the first approval you will need approval from the catalog maintainers as well. They will also be the ones to do the merge.

taraknath555 commented:
Hey @freben @awanlin just want to check about is this change is added in current release of 1.18.1 as after updating the backstage to same i.e. 1.18.1,  i am not able to take the advantage of the changes!

awanlin commented:
Hi @taraknath555, did you add the processor? This would be done in your `/packages/backend/src/plugins/catalog.ts` file and you would add this:

```ts
const integrations = ScmIntegrations.fromConfig(env.config);
 builder.addProcessor(new AnnotateScmSlugEntityProcessor(integrations))
  ```
  
  There'll be some imports as welll
